File: frontend/js/views/final_paycheck.js
// Final paycheck / waiting-time penalty generator — daily wage × days late
// (capped) plus unpaid final wages, via /calc/final-paycheck.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderFinalPaycheck(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.finalpay.h1.title">// FINAL PAYCHECK PENALTY</span></h1>
        <p class="muted small" data-i18n="view.finalpay.hint.intro">
            When an employer pays final wages late after a separation, many states impose a waiting-time
            penalty equal to the employee's daily wage for each day the payment is late, capped (California
            Labor Code §203 caps it at 30 days). It computes the daily wage, the capped penalty, and the total
            owed. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.finalpay.h2.inputs">Penalty inputs</h2>
            <form id="finalpay-form" class="inline-form">
                <label><span data-i18n="view.finalpay.label.state">State</span>
                    <input type="text" name="state" value="California" required></label>
                <label><span data-i18n="view.finalpay.label.employer">Employer</span>
                    <input type="text" name="employer_name" value=""></label>
                <label><span data-i18n="view.finalpay.label.employee">Employee</span>
                    <input type="text" name="employee_name" value=""></label>
                <label><span data-i18n="view.finalpay.label.wages">Unpaid final wages ($)</span>
                    <input type="number" step="50" min="0" name="final_wages_usd" value="1500"></label>
                <label><span data-i18n="view.finalpay.label.hourly">Hourly rate ($)</span>
                    <input type="number" step="0.01" min="0" name="hourly_rate_usd" value="25" required></label>
                <label><span data-i18n="view.finalpay.label.hpd">Hours per workday</span>
                    <input type="number" step="0.5" min="1" name="hours_per_day" value="8"></label>
                <label><span data-i18n="view.finalpay.label.dayslate">Days late</span>
                    <input type="number" step="1" min="0" name="days_late" value="10" required></label>
                <label><span data-i18n="view.finalpay.label.cap">Cap (days)</span>
                    <input type="number" step="1" min="0" name="cap_days" value="30"></label>
                <label><span data-i18n="view.finalpay.label.sepdate">Separation date</span>
                    <input type="date" name="separation_date" value="2026-06-01" required></label>
                <label><span data-i18n="view.finalpay.label.date">Statement date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.finalpay.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.finalpay.ph.statute'))}"></label>
            </form>
        </div>
        <div id="finalpay-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#finalpay-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            employer_name: (fd.get('employer_name') || '').trim(),
            employee_name: (fd.get('employee_name') || '').trim(),
            final_wages_usd: Number(fd.get('final_wages_usd')) || 0,
            hourly_rate_usd: Number(fd.get('hourly_rate_usd')) || 0,
            hours_per_day: Number(fd.get('hours_per_day')) || 8,
            days_late: Number(fd.get('days_late')) || 0,
            cap_days: Number(fd.get('cap_days')) || 30,
            separation_date: fd.get('separation_date'),
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcFinalPaycheck(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.finalpay.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const capNote = doc.cap_applied
        ? ` <span class="muted small" data-i18n="view.finalpay.capped">(capped)</span>`
        : '';
    const el = mount.querySelector('#finalpay-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.finalpay.card.total">Total owed</div>
                    <div class="value">${money(doc.total_owed_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.finalpay.card.penalty">Waiting-time penalty</div>
                    <div class="value">${money(doc.waiting_time_penalty_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.finalpay.card.days">Penalty days</div>
                    <div class="value">${doc.penalty_days}${capNote}</div></div>
                <div class="card"><div class="label" data-i18n="view.finalpay.card.daily">Daily wage</div>
                    <div class="value">${money(doc.daily_wage_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="finalpay-copy" type="button" data-i18n="view.finalpay.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="finalpay-download" type="button" data-i18n="view.finalpay.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#finalpay-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.finalpay.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.finalpay.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#finalpay-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'final-paycheck.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/final_paycheck.rs
//! Final paycheck / waiting-time penalty — when an employer pays final wages late
//! after a separation, many states impose a penalty equal to the employee's daily
//! wage for each day the payment is late, capped (California Labor Code §203 caps
//! it at 30 days). This computes the daily wage, the capped penalty days, the
//! waiting-time penalty, and the total owed (unpaid final wages plus penalty).
//! No existing generator computes a capped daily-wage penalty. Drafting aid, not
//! legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct FinalPaycheckInput {
    pub employer_name: String,
    pub employee_name: String,
    /// Unpaid final wages owed at separation.
    #[serde(default)]
    pub final_wages_usd: f64,
    /// Employee's hourly rate.
    pub hourly_rate_usd: f64,
    /// Hours in the employee's workday (for the daily wage).
    #[serde(default = "default_hours_per_day")]
    pub hours_per_day: f64,
    /// Calendar days the final payment is late.
    pub days_late: u32,
    /// Maximum penalty days (California §203 is 30).
    #[serde(default = "default_cap_days")]
    pub cap_days: u32,
    pub separation_date: String,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

fn default_hours_per_day() -> f64 {
    8.0
}

fn default_cap_days() -> u32 {
    30
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct FinalPaycheck {
    pub title: String,
    /// Hourly rate × hours per day.
    pub daily_wage_usd: f64,
    /// Days late, capped at cap_days.
    pub penalty_days: u32,
    /// True when the cap reduced the penalty days.
    pub cap_applied: bool,
    /// Daily wage × penalty days.
    pub waiting_time_penalty_usd: f64,
    pub final_wages_usd: f64,
    /// Final wages + penalty.
    pub total_owed_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &FinalPaycheckInput) -> FinalPaycheck {
    let daily = cents(i.hourly_rate_usd * i.hours_per_day);
    let penalty_days = i.days_late.min(i.cap_days);
    let cap_applied = i.days_late > i.cap_days;
    let penalty = cents(daily * penalty_days as f64);
    let total = cents(i.final_wages_usd + penalty);

    let cap_line = if cap_applied {
        format!(" The {} days late are capped at the {}-day maximum.", i.days_late, i.cap_days)
    } else {
        String::new()
    };

    let calc_body = format!(
        "The daily wage is {} ({} per hour × {} hours/day). The final payment is {} day(s) late.{} The waiting-time penalty is {} ({} × {} penalty day(s)). Added to unpaid final wages of {}, the total owed is {}.",
        money(daily),
        money(i.hourly_rate_usd),
        i.hours_per_day,
        i.days_late,
        cap_line,
        money(penalty),
        money(daily),
        penalty_days,
        money(i.final_wages_usd),
        money(total)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("The waiting-time penalty is imposed under the wage laws of the State of {}.", i.state)
    } else {
        format!("The waiting-time penalty is imposed under the wage laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Employer: {}\nEmployee: {}\nSeparation date: {}\nStatement date: {}",
                i.employer_name, i.employee_name, i.separation_date, i.date
            ),
        },
        DocClause {
            heading: "1. Final Wages".into(),
            body: format!(
                "Unpaid wages owed to the Employee at separation are {}. Final wages are due promptly after separation as required by state law.",
                money(i.final_wages_usd)
            ),
        },
        DocClause { heading: "2. Waiting-Time Penalty".into(), body: calc_body },
        DocClause {
            heading: "3. Total Owed".into(),
            body: format!(
                "The Employer owes the Employee {} in total: {} in unpaid final wages plus {} in waiting-time penalty.",
                money(total), money(i.final_wages_usd), money(penalty)
            ),
        },
        DocClause { heading: "4. Governing Law".into(), body: governing },
        DocClause {
            heading: "Certification".into(),
            body: format!(
                "Employer: ____________________  Date: __________\n{}",
                i.employer_name
            ),
        },
    ];

    FinalPaycheck {
        title: "Final Paycheck & Waiting-Time Penalty".into(),
        daily_wage_usd: daily,
        penalty_days,
        cap_applied,
        waiting_time_penalty_usd: penalty,
        final_wages_usd: cents(i.final_wages_usd),
        total_owed_usd: total,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> FinalPaycheckInput {
        FinalPaycheckInput {
            employer_name: "Acme Co".into(),
            employee_name: "Pat Former".into(),
            final_wages_usd: 1_500.0,
            hourly_rate_usd: 25.0,
            hours_per_day: 8.0,
            days_late: 10,
            cap_days: 30,
            separation_date: "2026-06-01".into(),
            date: "2026-07-01".into(),
            state: "California".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn penalty_within_cap() {
        let d = generate(&base());
        assert!(close(d.daily_wage_usd, 200.0));
        assert_eq!(d.penalty_days, 10);
        assert!(!d.cap_applied);
        assert!(close(d.waiting_time_penalty_usd, 2_000.0));
        assert!(close(d.total_owed_usd, 3_500.0));
    }

    #[test]
    fn penalty_capped_at_30_days() {
        let d = generate(&FinalPaycheckInput { days_late: 45, ..base() });
        assert_eq!(d.penalty_days, 30);
        assert!(d.cap_applied);
        assert!(close(d.waiting_time_penalty_usd, 6_000.0));
        assert!(close(d.total_owed_usd, 7_500.0));
    }

    #[test]
    fn on_time_no_penalty() {
        let d = generate(&FinalPaycheckInput { days_late: 0, ..base() });
        assert_eq!(d.penalty_days, 0);
        assert!(close(d.waiting_time_penalty_usd, 0.0));
        assert!(close(d.total_owed_usd, 1_500.0));
    }

    #[test]
    fn exactly_at_cap_not_flagged() {
        let d = generate(&FinalPaycheckInput { days_late: 30, ..base() });
        assert_eq!(d.penalty_days, 30);
        assert!(!d.cap_applied);
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&FinalPaycheckInput { statute_citation: "Cal. Lab. Code § 203".into(), ..base() });
        assert_eq!(d.statutory_citation, "Cal. Lab. Code § 203");
        assert!(d.clauses.iter().any(|c| c.body.contains("Cal. Lab. Code § 203")));
    }
}
